File: src/train_model.py
# src/train_model.py
import geopandas as gpd
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import joblib

def train_model(training_data, model_path):
    train_data = gpd.read_file(training_data)
    X = train_data[['nir', 'red']]
    y = train_data['class']
    
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
    
    clf = RandomForestClassifier(n_estimators=100, random_state=42)
    clf.fit(X_train, y_train)
    
    joblib.dump(clf, model_path)
    
    preds = clf.predict(X_test)
    accuracy = accuracy_score(y_test, preds)
    print(f"Model Accuracy: {accuracy:.2f}")

train_model("data/training_points.geojson", "models/rf_model.pkl")

File: src/train_model.R
# src/train_model.R
library(sf)
library(randomForest)

train_model <- function(training_data, model_path) {
  train_data <- st_read(training_data)
  X <- train_data[, c("nir", "red")]
  y <- as.factor(train_data$class)
  
  set.seed(42)
  train_idx <- sample(1:nrow(X), size = 0.8 * nrow(X))
  X_train <- X[train_idx, ]
  y_train <- y[train_idx]
  X_test <- X[-train_idx, ]
  y_test <- y[-train_idx]
  
  rf_model <- randomForest(X_train, y_train, ntree = 100, set.seed(42))
  saveRDS(rf_model, model_path)
  
  preds <- predict(rf_model, X_test)
  accuracy <- sum(preds == y_test) / length(y_test)
  print(paste("Model Accuracy:", round(accuracy, 2)))
}

train_model("data/training_points.geojson", "models/rf_model.rds")